cpp/src/arrow/table_test.cc

```diff
@@ -406,6 +407,165 @@ TEST_F(TestTable, ConcatenateTables) {
+      const std::string& column_name, const std::shared_ptr<DataType>& data_type,
+      const int length) {
+    std::shared_ptr<Array> array_of_nulls;
+    DCHECK_OK(MakeArrayOfNull(data_type, length, &array_of_nulls));
+    return Table::Make(schema({field(column_name, data_type)}), {array_of_nulls});
+  }
+};
+
+TEST_F(TestPromoteTableToSchema, IdenticalSchema) {
+  const int length = 10;
+  auto metadata =
+      std::shared_ptr<KeyValueMetadata>(new KeyValueMetadata({"foo"}, {"bar"}));
+  MakeExample1(length);
+  std::shared_ptr<Table> table = Table::Make(schema_, arrays_);
+
+  std::shared_ptr<Table> result;
+  ASSERT_OK(PromoteTableToSchema(table, schema_->WithMetadata(metadata),
+                                 default_memory_pool(), &result));
+
+  std::shared_ptr<Table> expected = table->ReplaceSchemaMetadata(metadata);
+
+  ASSERT_TRUE(result->Equals(*expected));
+}
+
+TEST_F(TestPromoteTableToSchema, PromoteNullTypeField) {
+  const int length = 10;
+  auto metadata =
+      std::shared_ptr<KeyValueMetadata>(new KeyValueMetadata({"foo"}, {"bar"}));
+  auto table_with_null_column = MakeTableWithOneNullFilledColumn("field", null(), length)
+                                    ->ReplaceSchemaMetadata(metadata);
+  auto promoted_schema = schema({field("field", int32())});
+
+  std::shared_ptr<Table> result;
+  ASSERT_OK(PromoteTableToSchema(table_with_null_column, promoted_schema,
+                                 default_memory_pool(), &result));
+
+  ASSERT_TRUE(
+      result->Equals(*MakeTableWithOneNullFilledColumn("field", int32(), length)));
+}
+
+TEST_F(TestPromoteTableToSchema, AddMissingField) {
+  const int length = 10;
+  auto f0 = field("f0", int32());
+  auto table = Table::Make(schema({}), std::vector<std::shared_ptr<Array>>(), length);
+  auto promoted_schema = schema({field("field", int32())});
+
+  std::shared_ptr<Table> result;
+  ASSERT_OK(PromoteTableToSchema(table, promoted_schema, default_memory_pool(), &result));
+
+  ASSERT_TRUE(
+      result->Equals(*MakeTableWithOneNullFilledColumn("field", int32(), length)));
+}
+
+TEST_F(TestPromoteTableToSchema, IncompatibleTypes) {
+  const int length = 10;
+  auto table = MakeTableWithOneNullFilledColumn("field", int32(), length);
+
+  std::shared_ptr<Table> result;
+  // Invalid promotion: int32 to null.
+  ASSERT_RAISES(Invalid, PromoteTableToSchema(table, schema({field("field", null())}),
+                                              default_memory_pool(), &result));
+
+  // Invalid promotion: int32 to uint32.
+  ASSERT_RAISES(Invalid, PromoteTableToSchema(table, schema({field("field", uint32())}),
+                                              default_memory_pool(), &result));
+}
+
+TEST_F(TestPromoteTableToSchema, DuplicateFieldNames) {
+  const int length = 10;
+
+  auto table = Table::Make(
+      schema({field("field", int32()), field("field", null())}),
+      {MakeRandomArray<Int32Array>(length), MakeRandomArray<NullArray>(length)});
```

Comment:
See also random array generation in https://github.com/apache/arrow/blob/master/cpp/src/arrow/testing/random.h

Reply:
Are you suggesting using them instead of MakeRandomArray<> (which is also a shared utility defined in testing/gtest_common.h )?

Reply:
It's fine as is, just mentioning it in case it makes things easier in the future